tokio/src/runtime/metrics/histogram.rs

```diff
@@ -35,31 +35,34 @@ pub(crate) struct HistogramBatch {
     resolution: u64,
 }
 
-cfg_unstable! {
-    /// Whether the histogram used to aggregate a metric uses a linear or
-    /// logarithmic scale.
-    #[derive(Debug, Copy, Clone, Eq, PartialEq)]
-    #[non_exhaustive]
-    pub enum HistogramScale {
-        /// Linear bucket scale
-        Linear,
-
-        /// Logarithmic bucket scale
-        Log,
-    }
+/// Whether the histogram used to aggregate a metric uses a linear or
+/// logarithmic scale.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[non_exhaustive]
+#[allow(unreachable_pub)]
+pub enum HistogramScale {
+    /// Linear bucket scale
+    Linear,
+
+    /// Logarithmic bucket scale
+    #[allow(dead_code)]
```

Comment:
is this change required for the mean time stabilization or just a coincidental change?

Reply:
I believe this change is required.

At master we have implemented another `struct WorkerMetrics` at `tokio/src/runtime/metrics/mock.rs` which didn't use `struct Histogram` defined at `tokio/src/runtime/metrics/histogram.rs`. 

In order to stabilize the API, I have to removed this mock WorkerMetrics and instead point to the "real" `WorkerMetrics` at `tokio/src/runtime/metrics/worker.rs`

The "real" `WorkerMetrics` has a field `poll_count_histogram` which is `Option<Histogram>`, and thus will attempt to parse `tokio/src/runtime/metrics/histogram.rs`. From there, you can see Histogram and HistogramBuilder both refers to `HistogramScale` which is hidden inside `tokio_unstable`. Clearly this wouldn't compile.

Reply:
I need to read through this PR in more detail so bare with me, but we shouldn't be making this stable and public if it isn't actually needed to to stabilize `worker_total_busy_duration` (and I'm mostly sure that it isn't).

In general, all the `#[allow(dead_code)]` that are required for this chnage give me the impression that we're exposing something that is only really used in `tokio_unstable` and so we should find a way to expose it only unstable `tokio_unstable`.

Reply:
Thanks for your comment!

I think the core issue here is that by exposing `worker_total_busy_duration`, we are also exposing the "real" `WorkerMetrics` at `tokio/src/runtime/metrics/worker.rs`, which in turn will attempt to parse `HistogramScale`.

Currently at master branch, when no `tokio_unstable` flag is passed in, the `WorkerMetrics` at `tokio/src/runtime/metrics/mock.rs` will be parsed (which is just an empty struct). And if the flag is passed in, the `WorkerMetrics` at `tokio/src/runtime/metrics/worker.rs` will be parsed

https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/metrics/mod.rs#L27-L40

Reply:
Yeah becoz `HistogramScale` is still behind `cfg_unstable_metrics`:

https://github.com/tokio-rs/tokio/pull/6899/files#diff-15ebfca6124144b3cffd9845fa3ce5596c342ce94d84dd406ee84f856db4c66cR23-R25

Reply:
OK, now I believe I understand.

I would suggest that instead of compiling with the entire worker metrics in order to access only the `busy_duration_total` field, we should gate all the fields that we won't be using on `cfg_unstable_metrics`. Otherwise users will still be paying the price for metrics which they don't have access to - and that is something that we would really like to avoid.

Have a look at the runtime Builder for an example of how we have fields and implementations that touch them gated on a cfg flag:

https://github.com/tokio-rs/tokio/blob/512e9decfb683d22f4a145459142542caa0894c9/tokio/src/runtime/builder.rs#L125-L126

As a general rule, if you need `#[allow(dead_code)]` then there is probably something that should be gated on a cfg flag instead.

Reply:
Thanks @hds . I've reverted changes to `histogram.rs`. Also the real `Histogram`, `HistogramBuilder` and `HistogramBatch` are gated behind unstable flag now, and instead I've created a mocked version of these for compilation. 

For `WorkerMetrics`,  as we target to stabilize more metrics, I'd suggest exposing all fields instead of stabilizing only `busy_duration_total` and putting other fields behind unstable.

Let me know what you think

Reply:
@Owen-CH-Leung the problem with stabilizing all of `WorkerMetrics` is that when `tokio_unstable` is **not** enabled, the runtime will pay the price for all those metrics, but there will be no way to access them. For this reason I think that it would be better to only stabilize what we're exposing.

Reply:
Thanks @hds . I've made changes to hide most of the fields of `WorkerMetrics` behind unstable flag, except for `busy_duration_total`, `queue_depth` and `thread_id`. The latter 2 fields are needed in order for `set_queue_depth` and `set_thread_id` to work properly.

I've also enriched the mock `MetricsBatch` to have minimal implementation of `batch::MetricsBatch` so that the `worker_total_busy_duration` API can function properly under stable build

Let me know your thoughts!